File: src/main/java/sda/studentmanagement/studentmanager/services/CourseService.java
package sda.studentmanagement.studentmanager.services;

import org.springframework.data.domain.Page;
import sda.studentmanagement.studentmanager.domain.Course;
import sda.studentmanagement.studentmanager.dto.AddCourseFormDto;
import sda.studentmanagement.studentmanager.projections.CourseDataChartsProjection;
import sda.studentmanagement.studentmanager.projections.CourseDataProjection;

import java.util.HashMap;
import java.util.List;

public interface CourseService {
    List<Course> getCourses();

    List<CourseDataChartsProjection> getCoursesCharts();

    List<HashMap<Object, Object>> getCourseCountCharts();

    Page<Course> getPaginatedCourses(int pageNumber, int pageSize);

    List<Course> getUserCourses(long userId, boolean userCourses);

    List<String> getUserCourseNames(long userId);

    List<CourseDataProjection> getCoursesData();

    Course getCourse(long courseId);

    void addUserToCourse(long userId, long courseId);

    void removeUserFromCourse(long userId, long courseId);

    void saveCourse(AddCourseFormDto addCourseForm);

    Course savePopulateCourse(Course course);

    Course editCourse(Course course);

    void deleteCourse(long courseId);
}

File: angular/studentmanagerapp/src/app/services/course.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { NotificationService } from './notification.service';
import { Course } from '../models/course.module';
import { ChartsInterface } from '../interfaces/charts-interface';

@Injectable({
  providedIn: 'root'
})
export class CourseService {
  headers = new HttpHeaders().set('Content-Type', 'application/json');
  private apiUrl: string = 'http://localhost:8080/api';

  constructor(private http: HttpClient, public notificationService: NotificationService) { }

  getCourses(): Observable<Course[]> {
    const url: string = `${this.apiUrl}/courses`;

    return this.http.get<Course[]>(url, { headers: this.headers });
  }

  getCourseUsersCharts(): Observable<ChartsInterface[]> {
    const url: string = `${this.apiUrl}/courses/data/charts`;

    return this.http.get<ChartsInterface[]>(url, { headers: this.headers });
  }

  getPaginatedCourses(pageNumber: number, pageSize: number): Observable<Course[]> {
    const url: string = `${this.apiUrl}/courses/paginated`;

    return this.http.get<Course[]>(url, {
      headers: this.headers, params: {
        pageNumber,
        pageSize,
      }
    });
  }

  getCourseCountCharts(): Observable<ChartsInterface> {
    const url: string = `${this.apiUrl}/courses/count/charts`;

    return this.http.get<ChartsInterface>(url, { headers: this.headers });
  }

  getCourse(courseId: number): Observable<Course> {
    const url: string = `${this.apiUrl}/course/${courseId}`;

    return this.http.get<Course>(url, { headers: this.headers });
  }

  getUserCourses(userId: number, checkCourseContainsUser: boolean): Observable<Course[]> {
    const url: string = `${this.apiUrl}/courses/user/${userId}?userCourses=${checkCourseContainsUser}`;

    return this.http.get<Course[]>(url, { headers: this.headers });
  }

  getCoursesData(): Observable<Course[]> {
    const url: string = `${this.apiUrl}/courses/data`;

    return this.http.get<Course[]>(url, { headers: this.headers })
  }

  joinCourse(userId: number, courseId: number) {
    const url: string = `${this.apiUrl}/course/add/user`;
    const body = { userId, courseId };

    return this.http.post(url, body, { responseType: 'text' });
  }

  addCourse(body: any): Observable<string> {
    const url: string = `${this.apiUrl}/course`;

    return this.http.post(url, body, { responseType: 'text' });
  }

  editCourse(courseId: number, body: any): Observable<string> {
    const url: string = `${this.apiUrl}/course/${courseId}`;

    return this.http.put(url, body, { responseType: 'text' });
  }

  leaveCourse(userId: number, courseId: number) {
    const url: string = `${this.apiUrl}/course/remove/user`;
    const body = { userId, courseId };

    return this.http.post(url, body, { responseType: 'text' });
  }

  deleteCourse(courseId: number): Observable<string> {
    const url: string = `${this.apiUrl}/course/${courseId}`;

    return this.http.delete(url, { responseType: 'text' });
  }
}
